### PR TITLE
adding parallel call and async invocation for AMS

### PR DESCRIPTION
**Pull request by elopezanaya**
# PR Details

### Description
_Include a description of the problem to be solved_

#### Main problem
Performance time for initialization are over 3 seconds, this is mainly due to AMS Client and sequencial calls when loding the componets, which doesnt have a dependency

## Solution Proposed

_Detail what is the solution proposed, include links to design document if required or any other document required to support the solution_

image

#### how to use it

```
const initializeOptionalParams: InitializeOptionalParams = {
        getLiveChatConfigOptionalParams: {
            useParallelLoad: true
        }
    };

 await chatSDK.initialize(initializeOptionalParams);
```

The component will have the option to load in parallel or in sequence, for those who doesnt want to change or load time is not a priority, but instead want all components fully loaded.

#### Parallel Load
Load AMS in the background and the rest of the components in parallel, with this change, load times are reduced to the range of 400ms to 700 ms, AMS component may not be fully loaded , but it contains a state machine that handles the states to ensure the component is ready when needed.

#### Sequencial Load
is the legacy implementation, for those that need the component fully loaded before start and performance is not an issue.
This will also serve as switch in case of a corner case no detected and its needed a fallback to legacy from paralllel to sequencial.

#### Other changes

- Fix multiple unit tests with false positives
- Add separate unit tests for parallel and sequential
- update pipeline to use node20

### Acceptance criteria

_Define what are the conditions to consider the PR has achieved the intended goal_

## Test cases and evidence

_Include what tests cases were considered, any evidence of testing for future references, to identify any corner cases, etc_

### OC-LCW Parallel load(~600ms)
image

### OC-LCW Sequential load (~4000ms)
image

### OOB Parallel load (~400ms)
image

### Sanity Tests

- [ ] You have tested all changes in Popout mode
- [ ] You have tested all changes in cross browsers i.e Edge, Chrome, Firefox, Safari and mobile devices(iOS and Android)
- [ ] Your changes are included in the [CHANGELOG](../CHANGE_LOG.md)

## A11y

- [ ] You have run the [Automated Accessibility Check](https://accessibilityinsights.io/docs/en/windows/getstarted/automatedchecks) and have no reported issues

***Please provide justification if any of the validations has been skipped.***

### CODE REVIEW

File: __tests__/OmnichannelChatSDK.spec.ts [hunk 1]
```diff
@@ -1507,10 +1312,10 @@ describe('Omnichannel Chat SDK', () => {
             }
         });
 
-        it('[LiveChatV1] ChatSDK.startChat() with customContext, browser, os, locale, device defined in sessionInitOptionalParams should pass it to OCClient.sessionInit() call\'s optional parameters', async() => {
```
Comment (xTEddie) [inline __tests__/OmnichannelChatSDK.spec.ts:1510]: This test is valid. Its goal is still validate the `customContext` are being sent to `sessioninit`. We need to remove the `v1` dependency on it

File: __tests__/OmnichannelChatSDK.spec.ts [hunk 2]
```diff
@@ -1536,25 +1332,16 @@ describe('Omnichannel Chat SDK', () => {
...
-            }
+            const chatContext = await chatSDK.getCurrentLiveChatContext();
 
-            expect(chatSDK.OCClient.sessionInit.mock.calls[0][1]).toMatchObject(sessionInitOptionalParams);
+            expect(Object.keys(chatContext).length).toBe(0);
+            expect(Object.keys(chatContext).includes('chatToken')).toBe(false);
+            expect(Object.keys(chatContext).includes('requestId')).toBe(false);
         });
 
-        it('[LiveChatV1] ChatSDK.startChat() with initContext defined should override IStartChatOptionalParams', async() => {
```
Comment (xTEddie) [inline __tests__/OmnichannelChatSDK.spec.ts:1554]: This test is valid. Its goal is validate the precedence of setting `initContext`. We need to remove the `v1` dependency

File: __tests__/OmnichannelChatSDK.spec.ts [hunk 3]
```diff
@@ -1563,64 +1350,53 @@ describe('Omnichannel Chat SDK', () => {
...
+                errorThrown = true;
             }
 
-            expect(chatSDK.OCClient.sessionInit.mock.calls[0][1]).toMatchObject(sessionInitOptionalParams);
+            expect(conversationDetails).toEqual({});
+            expect(chatSDK.OCClient.getLWIDetails).toHaveBeenCalledTimes(1);
+            expect(errorThrown).toBe(false);
         });
 
-        it('[LiveChatV1] ChatSDK.startChat() with authenticatedUserToken should pass it to OCClient.sessionInit() call\'s optional parameters', async() => {
```
Comment (xTEddie) [inline __tests__/OmnichannelChatSDK.spec.ts:1607]: This test is valid. Its goal is so validate that `authenticatedUserToken` is passed to `sessioninit`. We'd have to remove `v1` dependency on it

File: __tests__/OmnichannelChatSDK.spec.ts [hunk 4]
```diff
@@ -1643,270 +1415,37 @@ describe('Omnichannel Chat SDK', () => {
...
-                initContext: {}
-            }
+            await chatSDK.getConversationDetails();
 
-            expect(chatSDK.OCClient.sessionInit.mock.calls[0][1]).toMatchObject(sessionInitOptionalParams);
+            expect(chatSDK.OCClient.getLWIDetails).toHaveBeenCalledTimes(1);
+            expect(chatSDK.OCClient.getLWIDetails.mock.calls[0][1]).toMatchObject(optionalParams);
         });
 
-        it('[LiveChatV1] ChatSDK.getCallingToken() should return acs token if available', async () => {
```
Comment (xTEddie) [inline __tests__/OmnichannelChatSDK.spec.ts:1660]: Test is still valid. We need removal of `v1` dependency

File: src/OmnichannelChatSDK.ts [hunk 5]
```diff
@@ -194,66 +196,220 @@ class OmnichannelChatSDK {
...
+        // Constants for retry logic
+        const RETRY_DELAY_MS = 1000;
+        const MAX_RETRY_COUNT = 5;
+        let retryCount = 0;
+
+        while (retryCount < MAX_RETRY_COUNT) {
+            if (this.AMSClient) {
+                return this.AMSClient;
+            }
+            await new Promise(resolve => setTimeout(resolve, RETRY_DELAY_MS));
```
Comment (xTEddie) [inline src/OmnichannelChatSDK.ts:217]: There's a `sleep()` method that can be used in the `utils`

File: src/OmnichannelChatSDK.ts [hunk 6]
```diff
@@ -194,66 +196,220 @@ class OmnichannelChatSDK {
...
+            retryCount++;
+        }
 
-    public async initialize(optionalParams: InitializeOptionalParams = {}): Promise<ChatConfig> {
-        this.scenarioMarker.startScenario(TelemetryEvent.InitializeChatSDK);
+        console.error("AMSClient is not ready after retries");
```
Comment (xTEddie) [inline src/OmnichannelChatSDK.ts:221]: Should this be removed?

File: src/OmnichannelChatSDK.ts [hunk 7]
```diff
@@ -194,66 +196,220 @@ class OmnichannelChatSDK {
...
-            exceptionThrowers.throwUnsupportedLiveChatVersionFailure(new Error(ChatSDKErrorName.UnsupportedLiveChatVersion), this.scenarioMarker, TelemetryEvent.InitializeChatSDK);
+            exceptionThrowers.throwUnsupportedLiveChatVersionFailure(new Error(ChatSDKErrorName.UnsupportedLiveChatVersion), this.scenarioMarker, TelemetryEvent.InitializeComponents);
         }
 
+        this.isInitialized = true;
+        this.scenarioMarker.completeScenario(TelemetryEvent.InitializeComponents);
+    }
+
+    private async loadAmsClient() {
+        this.scenarioMarker.startScenario(TelemetryEvent.InitializeAMSClient);
```
Comment (xTEddie) [inline src/OmnichannelChatSDK.ts:262]: Would `InitializeMessagingClient` event make more sense? The method is initialization `ACSClient` & `AMSClient` which are part of the messaging platform we're using.

File: src/OmnichannelChatSDK.ts [hunk 8]
```diff
@@ -194,66 +196,220 @@ class OmnichannelChatSDK {
...
+
                 this.AMSClient = await createAMSClient({
                     framedMode: isBrowser(),
                     multiClient: true,
                     debug: false,
                     logger: this.amsClientLogger as PluggableLogger
                 });
-            } else if (this.liveChatVersion === LiveChatVersion.V1) {
+                this.AMSClientLoadCurrentState = AMSClientLoadStates.LOADED;
+            }else if (this.liveChatVersion === LiveChatVersion.V1) {
```
Comment (xTEddie) [inline src/OmnichannelChatSDK.ts:275]: nit: space after `}`

File: src/OmnichannelChatSDK.ts [hunk 9]
```diff
@@ -194,66 +196,220 @@ class OmnichannelChatSDK {
...
+     * @param optionalParams 
+     * @param parallel if true , it will run in parallel (fastest version) with components loaded in the background
+     * @returns livechatConfig
+     */
+    public async initialize(optionalParams: InitializeOptionalParams = {}): Promise<ChatConfig> {
+
+        const { getLiveChatConfigOptionalParams } = optionalParams;
+        const useParallelLoad = getLiveChatConfigOptionalParams?.useParallelLoad || false;
+        
+        if (useParallelLoad === true ){
```
Comment (xTEddie) [inline src/OmnichannelChatSDK.ts:384]: nit: space after `)`

File: src/OmnichannelChatSDK.ts [hunk 10]
```diff
@@ -194,66 +196,220 @@ class OmnichannelChatSDK {
...
+            return await this.parallelInitialization(optionalParams)
+        }
+           
+        return await this.sequentialInitialization(optionalParams);
+
```
Comment (xTEddie) [inline src/OmnichannelChatSDK.ts:389]: nit: remove space

File: __tests__/OmnichannelChatSDK.spec.ts [hunk 11]
```diff
@@ -490,20 +495,25 @@ describe('Omnichannel Chat SDK', () => {
...
             try {
                 await chatSDK.initialize();
+                fail();
             } catch (e) {
                 expect(e.message).toBe("OmnichannelClientInitializationFailure");
                 expect(chatSDK.OCClient).not.toBeDefined();
                 expect(chatSDK.getChatConfig).toHaveBeenCalledTimes(0);
             }
+
+            SDKProvider.getSDK= currentSDK;
```
Comment (xTEddie) [inline __tests__/OmnichannelChatSDK.spec.ts:516]: nit: space after `=`

File: src/OmnichannelChatSDK.ts [hunk 12]
```diff
@@ -194,66 +196,220 @@ class OmnichannelChatSDK {
...
+            this.scenarioMarker.startScenario(TelemetryEvent.InitializeChatSDKParallel);
+
+            if (this.isInitialized) {
+                this.scenarioMarker.completeScenario(TelemetryEvent.InitializeChatSDKParallel);
+                return this.liveChatConfig;
+            }
+    
+            this.useCoreServicesOrgUrlIfNotSet();
+
+            // these components will
```
Comment (xTEddie) [inline src/OmnichannelChatSDK.ts:298]: Is the comment needed?

File: src/core/GetLiveChatConfigOptionalParams.ts [hunk 13]
```diff
@@ -1,6 +1,7 @@
 interface GetLiveChatConfigOptionalParams {
     sendCacheHeaders?: boolean;
     useRuntimeCache?: boolean;
+    useParallelLoad?: boolean;
```
Comment (xTEddie) [inline src/core/GetLiveChatConfigOptionalParams.ts:4]: Would it make more sense if  `useParallelLoad` was  part of `InitializeOptionalParams` instead?

GetLiveChatConfigOptionalParams is the params used specifically for getting live chat config. 

`sendCacheHeaders` is applied upon getting live chat config call
`useRuntimeCache` is applied to retrieve the chat config from the `ChatSDK` instance 

`useParallelLoad` would make more sense if it in the `initialization` context since it's not for getting live chat config specifically.

Comment (xTEddie) [reply src/core/GetLiveChatConfigOptionalParams.ts:4]: Would it make sense to use `useConcurrency` instead? `Promise.all()` technically execute the promises concurrently and not parallelly

File: src/telemetry/TelemetryEvent.ts [hunk 14]
```diff
@@ -1,7 +1,11 @@
 enum TelemetryEvent {
     ValidateOmnichannelConfig = "ValidateOmnichannelConfig",
     ValidateSDKConfig = "ValidateSDKConfig",
-    InitializeChatSDK = "InitializeChatSDK",
+    InitializeChatSDKSequential = "InitializeChatSDKSequential",
```
Comment (xTEddie) [inline src/telemetry/TelemetryEvent.ts:4]: I see that we introduced two new events on initialization: `InitializeChatSDKSequential` & `InitializeChatSDKParallel`. 

Previously, we'd only have `InitializeChatSDK`. Is there any monitors that are dependent on `InitializeChatSDK` events? If monitors or other components dependent on this event would potentially stop working.

In order to keep `back comp`, keeping `InitializeChatSDK` for the sequential flow and using `InitializeChatSDKUsingConcurrency` for the "concurrent" flow may be safer.

File: src/core/ChatSDKError.ts [hunk 15]
```diff
@@ -53,7 +53,10 @@ export enum ChatSDKErrorName {
     /** Failure on retrieving AuthToken from AuthTokenProvider */
     GetAuthTokenFailed = "GetAuthTokenFailed",
     /** AuthToken is undefined */
-    UndefinedAuthToken = "UndefinedAuthToken"
+    UndefinedAuthToken = "UndefinedAuthToken",
+
+    UnknownAMSLoadState = "UnknownAMSLoadState",
```
Comment (xTEddie) [inline src/core/ChatSDKError.ts:58]: `UnknownAttachmentClientLoadState`?

File: src/telemetry/TelemetryEvent.ts [hunk 16]
```diff
@@ -1,7 +1,11 @@
...
+    InitializeChatSDKParallel = "InitializeChatSDKParallel",
+    InitializeAMSClient= "InitializeAMSClient",
```
Comment (xTEddie) [inline src/telemetry/TelemetryEvent.ts:6]: nit: space before `=`

File: src/OmnichannelChatSDK.ts (hunk 6 above)
Comment (elopezanaya) [reply src/OmnichannelChatSDK.ts:221]: done

File: src/core/GetLiveChatConfigOptionalParams.ts (hunk 13 above)
Comment (elopezanaya) [reply src/core/GetLiveChatConfigOptionalParams.ts:4]: image

**Review by mithunjohny — APPROVED** [review]

File: src/OmnichannelChatSDK.ts [hunk 17]
```diff
@@ -194,66 +196,214 @@ class OmnichannelChatSDK {
...
+
         this.debug = flag;
-        this.AMSClient?.setDebug(flag);
+        this.getAMSClient().then((client) => client?.setDebug(flag));
+
         this.telemetry?.setDebug(flag);
         this.scenarioMarker.setDebug(flag);
         loggerUtils.setDebug(flag, this.ocSdkLogger, this.acsClientLogger, this.acsAdapterLogger, this.callingSdkLogger, this.amsClientLogger, this.ic3ClientLogger);
     }
+    private async retryLoadAMSClient() {
```
Comment (xTEddie) [inline src/OmnichannelChatSDK.ts:207]: nit: missing new line before `retryLoadAMSClient()`

**Review by xTEddie — APPROVED** [review]
